### PR TITLE
build: build proper VMDK formats for ESXI 6.5

### PR DESCRIPTION
**Pull request by jsommerville-untangle**
VMWare ESXI 6.5 and above is not compatible with
subformat=monolithicSparse (The default qemu-img convert -O VMDK option).
Monolithic Sparse vmdk can be imported, but issues occur when running
sysupgrade with new images and other tasks that modify the file system
(issues like Kernel panics, reboot loops, sometimes crashing the Host ESXI
box).

This change creates an additional VMDK output file for ESXI that sets the
subformat to monlithicFlat, and the adapter_type to the SCSI lsilogic
controller.

This change existed back on:
25e36d3

But it looks like the change was removed when refactoring occurred with:
5f6a273

Signed-off-by: John Sommerville jsommerville@untangle.com

### CODE REVIEW

**Comment by feckert** [comment]
Did you send the patches for qemu-img upstream?
Or are this backports from qemu?
The version for qemu-img in openwrt is very old https://github.com/openwrt/openwrt/blob/65361171f8f1cc06db4ac1f7612090de8e697f52/tools/qemu/Makefile#L11

Upstream  version is now https://github.com/qemu/qemu/tree/stable-3.1

**Comment by jsommerville-untangle** [comment]
@feckert @bmastbergen backported the changes needed from future qemu versions

File: target/linux/x86/image/Makefile [hunk 1]
```diff
@@ -134,6 +134,12 @@ ifneq ($(CONFIG_VMDK_IMAGES),)
 	$(STAGING_DIR_HOST)/bin/qemu-img convert -f raw -O vmdk \
 		$(BIN_DIR)/$(IMG_COMBINED)-$(1).img \
 		$(BIN_DIR)/$(IMG_COMBINED)-$(1).vmdk
+	#build the ESXI VMDK with 
+	rm $(BIN_DIR)/$(IMG_COMBINED)-$(1)-esxi.vmdk || true
+	rm $(BIN_DIR)/$(IMG_COMBINED)-$(1)-esxi-flat.vmdk || true
```
Comment (hauke) [inline target/linux/x86/image/Makefile]: Why do you delete this file?

**Comment by hauke** [comment]
The backport commit has format issues, please follow this: https://openwrt.org/submitting-patches

File: target/linux/x86/image/Makefile (hunk 1 above)
Comment (jsommerville-untangle) [reply target/linux/x86/image/Makefile]: This is following the pattern of the other image Build definitions to remove the output files before attempting the qemu-img convert process. The monolithic flat vmdk disk type requires a descriptor file (*-esxi.vmdk) as well as the flat image file (*-esxi-flat.vmdk). Both of these files must be loaded into the ESXI data store for a valid VMDK image to be available.

Comment (hauke) [reply target/linux/x86/image/Makefile]: where is the *-esxi-flat.vmdk file created?

Comment (jsommerville-untangle) [reply target/linux/x86/image/Makefile]: The output of qemu-img convert to monolithicFlat produces both the descriptor file (-esxi.vmdk) and the image file (-esxi-flat.vmdk).

**Comment by jsommerville-untangle** [comment]
@hauke We believe the format issue involved the backport commit missing the sign off, just force pushed the branch with these changes included.

Thanks!

**Comment by hauke** [comment]
Thank you for your pull request, it was applied to master.

**Comment by BartKoppers** [comment]
Is the esxi-build working at all?  I expected some output in https://downloads.openwrt.org/releases/19.07.2/targets/x86/64/

FWIW: 
To get an usable VMDK for ESXi 6.5, I had to manually include the  *-o adapter_type=lsilogic,subformat=monolithicFlat* option, so 
`qemu-img convert -f raw -O vmdk -o adapter_type=lsilogic,subformat=monolithicFlat ~/Downloads/openwrt-19.07.2-x86-64-combined-ext4.img openwrt-esxi.vmdk`

- qemu-img version 4.2.0 on MacOS
- didn't test on VMWare Workstation

**Pull request closed without merging**